core-bundle/src/EventListener/PageAccessListener.php

```diff
@@ -0,0 +1,97 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\EventListener;
+
+use Contao\CoreBundle\Exception\AccessDeniedException;
+use Contao\CoreBundle\Exception\InsufficientAuthenticationException;
+use Contao\CoreBundle\Framework\ContaoFramework;
+use Contao\CoreBundle\Security\ContaoCorePermissions;
+use Contao\PageModel;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Event\RequestEvent;
+use Symfony\Component\Security\Core\Security;
+
+/**
+ * @internal
+ */
+class PageAccessListener
+{
+    private ContaoFramework $framework;
+    private Security $security;
+
+    public function __construct(ContaoFramework $framework, Security $security)
+    {
+        $this->security = $security;
+        $this->framework = $framework;
+    }
+
+    /**
+     * Resolve pageModel in request attributes and check permission to access the page.
+     */
+    public function __invoke(RequestEvent $event): void
+    {
+        $request = $event->getRequest();
+        $pageModel = $this->getPageModel($request);
+
+        if (null === $pageModel) {
+            return;
+        }
+
+        $pageModel->loadDetails();
+        $request->attributes->set('pageModel', $pageModel);
+
+        if (!$pageModel->protected) {
+            return;
+        }
+
+        // Do not check for logged in member if -1 (guest group) is allowed
+        if (
+            !\in_array(-1, array_map('intval', $pageModel->groups), true)
+            && !$this->security->isGranted('ROLE_MEMBER')
+        ) {
+            throw new InsufficientAuthenticationException('Not authenticated');
+        }
+
+        if (!$this->security->isGranted(ContaoCorePermissions::MEMBER_IN_GROUPS, $pageModel->groups)) {
+            throw new AccessDeniedException('Member does not have access to page ID '.$pageModel->id);
+        }
+    }
+
+    private function getPageModel(Request $request): ?PageModel
+    {
+        if (!$request->attributes->has('pageModel')) {
+            return null;
+        }
+
+        $pageModel = $request->attributes->get('pageModel');
+
+        if (
+            isset($GLOBALS['objPage'])
+            && $GLOBALS['objPage'] instanceof PageModel
+            && (
+                ($pageModel instanceof PageModel && (int) $pageModel->id === (int) $GLOBALS['objPage']->id)
+                || (!$pageModel instanceof PageModel && (int) $GLOBALS['objPage']->id === (int) $pageModel)
+            )
+        ) {
+            return $GLOBALS['objPage'];
+        }
+
+        if ($pageModel instanceof PageModel) {
+            return $pageModel;
+        }
+
+        $this->framework->initialize();
+
+        return $this->framework->getAdapter(PageModel::class)->findByPk((int) $pageModel);
```

Comment:
Btw. the original logic in `AbstractFragmentController` and `ContaoContext` is a little different:

If no request is available or no `pageModel` attribute is available, then `$GLOBALS['objPage']` will be returned, if available (otherwise null).

The new version now always returns null, if no `pageModel` attribute is available. Not sure if that makes a significant difference though.

Reply:
good point. The _request's_ page model is actually empty. Maybe you should get the (global) page model from the master request if you want that?